backends/concrete-cpu/implementation/src/implementation/fast_keyswitch.rs

```diff
@@ -0,0 +1,272 @@
+
+    let (glwe_common_mask, additional_mask) = glwe_full_mask.split_at_mut(input.lwe_dimension);
+
+    let (lwe_body, lwe_mask) = input.into_data().split_last().unwrap();
+
+    // We copy the mask (each polynomial is in the wrong order)
+    glwe_common_mask.copy_from_slice(lwe_mask);
+
+    additional_mask.fill(0);
+
+    glwe_body.fill(0);
+
+    // We copy the body
+    glwe_body[0] = *lwe_body;
+
+    // We compute the number of elements which must be
+    // turned into their opposite
+    let opposite_count = polynomial_size - 1;
+
+    // We loop through the polynomials (as mut tensors)
+    for glwe_mask_poly in glwe_full_mask.chunks_exact_mut(polynomial_size) {
+        // We rotate the polynomial properly
+        glwe_mask_poly.rotate_right(opposite_count);
+
+        // We compute the opposite of the proper coefficients
+        for x in glwe_mask_poly[0..opposite_count].iter_mut() {
+            *x = x.wrapping_neg()
+        }
+
+        // We reverse the polynomial
+        glwe_mask_poly.reverse();
+    }
+}
+
+/// Performs the external product of `ggsw` and `glwe`, and stores the result in `out`.
+pub fn fast_keyswitch(
+    out: LweCiphertext<&mut [u64]>,
+    ggsw: GgswCiphertext<&[f64]>,
+    input: LweCiphertext<&[u64]>,
+    fft: FftView<'_>,
+    stack: DynStack<'_>,
+) -> (Vec<u64>, Vec<u64>) {
+    let mut input_glwe_c = vec![0; (ggsw.in_glwe_dimension + 1) * ggsw.polynomial_size];
+
+    let mut input_glwe = GlweCiphertext::new(input_glwe_c.as_mut_slice(), ggsw.in_glwe_params());
+
+    let mut output_glwe_c = vec![0; (ggsw.out_glwe_dimension + 1) * ggsw.polynomial_size];
+
+    let mut output_glwe = GlweCiphertext::new(output_glwe_c.as_mut_slice(), ggsw.out_glwe_params());
+
+    trivial_packing(input_glwe.as_mut_view(), input);
+
+    external_product(
+        output_glwe.as_mut_view(),
+        ggsw,
+        input_glwe.as_view(),
+        fft,
+        stack,
+    );
+
+    output_glwe.as_view().sample_extract(out, 0);
+
+    (input_glwe_c, output_glwe_c)
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::c_api::types::tests::to_generic;
+    use crate::implementation::encrypt::encrypt_constant_ggsw;
+    use crate::implementation::fast_keyswitch::fast_keyswitch;
+    use crate::implementation::fft::Fft;
+    use crate::implementation::types::{
+        CsprngMut, DecompParams, GgswCiphertext, GlweCiphertext, GlweParams, GlweSecretKey,
+        LweCiphertext, LweSecretKey,
+    };
+    use concrete_csprng::generators::{RandomGenerator, SoftwareRandomGenerator};
+    use concrete_csprng::seeders::Seed;
+    use dyn_stack::DynStack;
+    use std::mem::MaybeUninit;
+
+    use super::trivial_packing;
+
+    fn init_sk(sk: &mut [u64], csprng: CsprngMut) {
+        let len = sk.len();
+        LweSecretKey::new(sk, len).fill_with_new_key(csprng)
+    }
+
+    #[test]
+    fn test_trivial_packing() {
+        let in_lwe_dim = 100;
```

Comment:
It's better to take realistic parameters, for instance `in_lwe_dim=650`

Reply:
Done